executor/read/pattern_executor.rs

```diff
@@ -0,0 +1,188 @@
+
+use crate::{
+    batch::{FixedBatch, FixedBatchRowIterator},
+    error::ReadExecutionError,
+    pipeline::stage::ExecutionContext,
+    read::{
+        collecting_stage_executor::CollectedStageIterator, nested_pattern_executor::SubQueryResult,
+        step_executor::StepExecutors,
+    },
+    ExecutionInterrupt,
+};
+
+#[derive(Clone)]
+pub(super) struct BranchIndex(pub usize);
+
+#[derive(Clone)]
+pub(super) struct InstructionIndex(pub usize);
+
+impl InstructionIndex {
+    fn next(&self) -> InstructionIndex {
+        InstructionIndex(self.0 + 1)
+    }
+}
+
+pub(super) enum ControlInstruction {
+    Start(FixedBatch),
+
+    ExecuteImmediate(InstructionIndex),
+
+    MapRowBatchToRowForSubQuery(InstructionIndex, FixedBatchRowIterator),
+    ExecuteSubQuery(InstructionIndex, BranchIndex),
+
+    CollectingStage(InstructionIndex),
+    StreamCollected(InstructionIndex, CollectedStageIterator),
+
+    Yield(FixedBatch),
+}
+
+pub(crate) struct PatternExecutor {
+    executors: Vec<StepExecutors>,
+    control_stack: Vec<ControlInstruction>,
+}
+
+impl PatternExecutor {
+    pub(crate) fn new(executors: Vec<StepExecutors>) -> Self {
+        PatternExecutor { executors, control_stack: Vec::new() }
+    }
+
+    pub(crate) fn compute_next_batch(
+        &mut self,
+        context: &ExecutionContext<impl ReadableSnapshot + 'static>,
+        interrupt: &mut ExecutionInterrupt,
+    ) -> Result<Option<FixedBatch>, ReadExecutionError> {
+        self.batch_continue(context, interrupt)
+    }
+
+    pub(crate) fn prepare(&mut self, input_batch: FixedBatch) {
+        debug_assert!(self.control_stack.is_empty());
+        self.reset();
+        self.control_stack.push(ControlInstruction::Start(input_batch));
+    }
+
+    pub(super) fn reset(&mut self) {
+        self.control_stack.clear();
+    }
+
+    pub(super) fn batch_continue(
+        &mut self,
+        context: &ExecutionContext<impl ReadableSnapshot + 'static>,
+        interrupt: &mut ExecutionInterrupt,
+    ) -> Result<Option<FixedBatch>, ReadExecutionError> {
+        if self.control_stack.len() == 0 {
+            debug_assert!(self.control_stack.len() > 0);
+        }
+        while self.control_stack.last().is_some() {
+            let Self { control_stack, executors } = self;
+            // TODO: inject interrupt into Checkers that could filter out many rows without ending as well.
+            if let Some(interrupt) = interrupt.check() {
+                return Err(ReadExecutionError::Interrupted { interrupt });
+            }
+            let mut step = control_stack.pop().unwrap();
+
+            match step {
+                ControlInstruction::Start(batch) => {
+                    self.prepare_next_instruction_and_push_to_stack(context, InstructionIndex(0), batch)?;
+                }
+                ControlInstruction::ExecuteImmediate(index) => {
+                    let found = executors[index.0].unwrap_immediate().batch_continue(context, interrupt)?;
```

Comment:
Seomtimes just to avoid coming up with an arbitrary name i'll split this kind of code into two different lines:
```
let executor = executors[index.0].unwrap_immediate();
if let Some(bach) = executor.batch_continue(context, interrupt)?;
```

In this case I think it is much more readable!